dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+        """
+        return _bin_normalize(x, self.mmin, self.mdelta)
+
+
+    def bin_index(self, xnorm):
+        r""" Bin index associated to the given normalized intensity
+
+        The return value is an integer in [padding, nbins - 1 - padding]
+
+        Parameters
+        ----------
+        xnorm : float
+            intensity value normalized to the range covered by the histogram
+
+        Returns
+        -------
+        bin : int
+            the bin index associated to the given normalized intensity
+        """
+        return _bin_index(xnorm, self.nbins, self.padding)
+
+
+    def update_pdfs_dense(self, static, moving, smask=None, mmask=None):
+        r''' Computes the Joint Probability Density Function of two images
+
+        Parameters
+        ----------
+        static: array, shape (S, R, C)
+            static image
+        moving: array, shape (S, R, C)
+            moving image
+        smask: array, shape (S, R, C)
+            mask of static object being registered (a binary array with 1's
+            inside the object of interest and 0's along the background).
+            If None, ones_like(static) is used as mask.
+        mmask: array, shape (S, R, C)
+            mask of moving object being registered (a binary array with 1's
+            inside the object of interest and 0's along the background).
+            If None, ones_like(moving) is used as mask.
+        '''
+        dim = len(static.shape)
+        if dim == 2:
+            _compute_pdfs_dense_2d(static, moving, smask, mmask, self.smin,
+                                   self.sdelta, self.mmin, self.mdelta,
+                                   self.nbins, self.padding, self.joint,
+                                   self.smarginal, self.mmarginal)
+        elif dim == 3:
+            _compute_pdfs_dense_3d(static, moving, smask, mmask, self.smin,
+                                   self.sdelta, self.mmin, self.mdelta,
+                                   self.nbins, self.padding, self.joint,
+                                   self.smarginal, self.mmarginal)
+        else:
+            msg = 'Only dimensions 2 and 3 are supported. '+str(dim)+' received'
+            raise ValueError(msg)
+
+
+    def update_pdfs_sparse(self, sval, mval):
+        r''' Computes the Probability Density Functions of paired intensities
+
+        Parameters
+        ----------
+        sval: array, shape (n,)
+            sampled intensities from the static image at sampled_points
+        mval: array, shape (n,)
+            sampled intensities from the moving image at sampled_points
+        '''
+        energy = _compute_pdfs_sparse(sval, mval, self.smin, self.sdelta,
+                                      self.mmin, self.mdelta, self.nbins,
+                                      self.padding, self.joint, self.smarginal,
+                                      self.mmarginal)
+
+
+    def update_gradient_dense(self, theta, transform, static, moving,
```

Comment:
That's right, it does the full computation. Do you think something like "recompute" would be more clear?
Yes, it goes to self.joint_grad, I added a comment about that. =)